### ISSUE TITLE
Detect read-only not private fields and fix adding the "readonly" modifier

### ISSUE THREAD

**Issue by giggio** [code-cracker/code-cracker#177]
Related to #86 which is already solved and is specific to **private** fields. This one is specific to non private fields.

We have to verify that the field is not set according to the accessibility of field before offering the diagnostic: if it public in the whole solution, if it is protected on the whole solution only on inherited classes, internal only on the project and protected internal only on the inherited classes on the project.
In no case should the diagnostic be reported if it would break the code when the code fix (making the field read only) is applied. 

This:

````csharp
class Foo
{
    public int a = 1;
}
````

Becomes:

````csharp
class Foo
{
    public readonly int a = 1;
}
````

See more test cases on test class `CodeCracker.Tests.Usage.ReadonlyFieldTests`.

This is only an analyzer, the code fix is already written for private fields and should work. This is also the reason we are using the same diagnostic id.

This is an `Info`.
Category is `Usage`
Diagnostic id: `CC0074`

@dmgandini is working on it.

**Comment by dmgandini** [comment]
I believe the second is most common scenario as well. I want to work on it.

**Comment by giggio** [comment]
@dmgandini Ok, it is yours.
Only to let you know, this is a hard one. You **cannot** `.Wait` or `.Result` on an analyzer, ok?
The Roslyn team has given us some clues on how to work on this one. You have to register an analysis on compilation start, collect all the references, then on compilation end you verify if the field was ever written to. So this will only run on full compilations, not when the user is typing. This seems to be a limitation on this specific analysis that we want to do.
I will mark it as working.

**Comment by dmgandini** [comment]
I'm aware that is hard one. I was thinking to use SymbolFinder after load the current workspace, I was trying to figure out how to do it without use the MSBuildWorkspace once the project is a portable one. 
After your post I will try to do as you said. Thanks for the tips. If I get stuck I will let you guys know.

The diagnostic id, CC0052, is the same of ReadonlyFieldAnalyzer. Is it Ok?

**Comment by giggio** [comment]
Don't use the Workspace, as it is not portable, and I don't think it will ever be portable.
Yes, same diagnostic id, it is pretty much the same diagnostic, just a different scope. We could add a new one, but I am not sure if it makes sense. What do you think?

**Comment by giggio** [comment]
@dmgandini How is this coming? Need help?

**Comment by dmgandini** [comment]
Hi @giggio!
I had some personal issues that drained my time, although I'm still working on it.
Probably a PR will be send until weekend, 'cause the code is finished and I'm reviewing the code in order to check if exists any bug.

Regarding the diagnostic ID, it is a different scope and I would like to have a chance to download it or not. It is not a common scenario to have public/internal/protected fields exposed by a type, then it will not be useful to everyone. Make a long story short, once the #47 is done, it will be nice to have two different diagnostics.
What do you think?

**Comment by giggio** [comment]
Fine by me. Just update the diagnostic id on the wiki.
And note that we updated the project to VS 2015 CTP 6, you will need to update your code as well.

**Comment by dmgandini** [comment]
Please check the #292.
By the way, the automatic build is amazing! :-)
